python/preprocessor.py

```diff
@@ -0,0 +1,349 @@
+
+        returns
+        -------
+        preprocessed feature array
+        """
+        # convert feature names to observable names
+        observables = np.array([self.observable_name_dict[feature] for feature in features])
+
+        # use as a checklist to mark the items that are done
+        task_list = (self.config[FEATURE]).copy()
+
+        logger.debug("Observable order before preprocessing: "+str(observables))
+
+        while task_list:
+            function_name, list_of_observable = task_list.popitem(last = False)
+            logger.debug("Applying preprocessing function " + function_name)
+            function = self.feature_preprocessing_function_map[function_name]
+
+            # create a mask for the next operation
+            mask = self._mask(observables, list_of_observable)
+
+            # call preprocessor function, passing any additional args as is
+            # modify args here to add in additional arguments passed to preprocessor function
+            feature_array, observables = function(feature_array, mask, observables, **args)
+
+            logger.debug("Observable order after preprocessing round: "+str(observables))
+
+        gc.collect()
+        
+        # return the feature array after preprocessing
+        return feature_array
+    
+    def preprocess_weight(self, feature_arrays, weights, observables, **args):
+        """
+        apply the list of weight preprocessors in sequence and return the final set of preprocessed weights
+
+        arguments
+        ---------
+        feature_arrays: tuple of numpy arrays
+            feature arrays in the order of [data, sim, gen]
+        weights: tuple of numpy arrays
+            event weights in the same order as feature arrays
+        observables: list of str
+            indicating the position of each observable in feature array
+        """
+
+        for function in self.weight_preprocessing_functions:
+            weights = function(feature_arrays, weights, observables, **args)
+
+        return weights
+
+    def use_utility(self):
+        """
+        returns
+        -------
+        whether the config specified the utility functions to be used
+        """
+        return self.utility
+
+    # other functions that can be called directly as an utility
+
+    def standardize(self, feature_array):
+        """
+        standardize the given feature array
+
+        arguments
+        ---------
+        feature_arrays: numpy 2d arrays of shape (number of events, features in each event)
+
+        returns
+        -------
+        feature_array standardized to a mean of 0 and standard deviation of one
+        """
+
+        # compute mean over all feature_arrays together
+        mean = np.mean(feature_array, axis=0)
+        std = np.std(feature_array, axis=0)
+
+        return (feature_array - mean) / std
+
+    def group_standardize(self, feature_arrays):
```

Comment:
I changed group_standardize to using looping method. I'll make sure it works and do the same for the other functions.

Reply:
Thanks! I was also testing some implementation for computing this by hand. I'll just put it here in case it is useful.
```
def group_standardize(feature_arrays):

    s, s2, N = 0, 0, 0

    for farr in feature_arrays:
        s += np.sum(farr,axis=0)
        s2 += np.sum(farr*farr,axis=0)
        N += len(farr)

    mean = s / N
    std  = np.sqrt( s2/N - mean*mean )

    feature_arrays = ((feature_array - mean) / std for feature_array in feature_arrays)

    return feature_arrays
```

Reply:
I see. I'll look into which one is more efficient. I think in general the <s^2> <s>^2 method is more efficient.